packages/node/src/indexer/unfinalizedBlocks.service.ts

```diff
@@ -0,0 +1,222 @@
+    return this.apiService.getApi();
+  }
+
+  private get finalizedBlockNumber(): number {
+    return this.finalizedBlock.block.header.number.toNumber();
+  }
+
+  private getSortedUnfinalizedBlocks(): [string, HexString][] {
+    return Object.entries(this.unfinalizedBlocks).sort(
+      ([bestBlockA], [bestBlockB]) => {
+        return Number(bestBlockB) - Number(bestBlockA);
+      },
+    );
+  }
+
+  async processUnfinalizedBlocks(
+    block: SubstrateBlock,
+    tx: Transaction,
+  ): Promise<number | null> {
+    await this.registerUnfinalizedBlock(
+      block.block.header.number.toNumber(),
+      block.hash.toHex(),
+      tx,
+    );
+    if (!(await this.hasForked())) {
+      // Remove blocks that are now confirmed finalized
+      await this.deleteFinalizedBlock(tx);
+    } else {
+      // Get the last unfinalized block that is now finalized
+      return this.getLastCorrectFinalizedBlock();
+    }
+
+    return null;
+  }
+
+  registerFinalizedBlock(block: SignedBlock): void {
+    if (
+      this.finalizedBlock &&
+      this.finalizedBlockNumber >= block.block.header.number.toNumber()
+    ) {
+      return;
+    }
+    this.finalizedBlock = block;
+  }
+
+  private async registerUnfinalizedBlock(
+    blockNumber: number,
+    hash: HexString,
+    tx: Transaction,
+  ): Promise<void> {
+    if (blockNumber > this.finalizedBlockNumber) {
+      this.unfinalizedBlocks[blockNumber] = hash;
+      await this.saveUnfinalizedBlocks(this.unfinalizedBlocks, tx);
+    }
+  }
+
+  private async deleteFinalizedBlock(tx: Transaction): Promise<void> {
+    if (
+      this.lastCheckedBlockHeight !== undefined &&
+      this.lastCheckedBlockHeight < this.finalizedBlockNumber
+    ) {
+      this.removeFinalized(this.finalizedBlockNumber);
+      await this.saveLastFinalizedVerifiedHeight(this.finalizedBlockNumber, tx);
+      await this.saveUnfinalizedBlocks(this.unfinalizedBlocks, tx);
+    }
+    this.lastCheckedBlockHeight = this.finalizedBlockNumber;
+  }
+
+  // remove any records less and equal than input finalized blockHeight
+  private removeFinalized(blockHeight: number): void {
+    Object.entries(this.unfinalizedBlocks).map(([bestBlockHeight, hash]) => {
+      if (Number(bestBlockHeight) <= blockHeight) {
+        delete this.unfinalizedBlocks[bestBlockHeight];
+      }
+    });
+  }
+
+  // find closest record from block heights
+  private getClosestRecord(
+    blockHeight: number,
+  ): { blockHeight: number; hash: HexString } | undefined {
+    // Have the block in the best block, can be verified
+    const record = this.getSortedUnfinalizedBlocks().find(
+      ([bestBlockHeight, hash]) => Number(bestBlockHeight) <= blockHeight,
+    );
+    if (record) {
+      const [bestBlockHeight, hash] = record;
+      return { blockHeight: Number(bestBlockHeight), hash };
+    }
+    return undefined;
+  }
+
+  // check unfinalized blocks for a fork
+  private async hasForked(): Promise<boolean> {
```

Comment:
I feel `hasForked()` and `getLastCorrectFinalizedBlock()` can be merged. Do both in once iteration

Reply:
```suggestion
  private async hasForked(): Promise<BlockNumber> {
```

Reply:
I'm not sure we gain anything from this. Its easy to detect a fork because block hashes don't match, it takes more work to find at what block the fork occurred

Reply:
> I'm not sure we gain anything from this. Its easy to detect a fork because block hashes don't match, it takes more work to find at what block the fork occurred

But you still need to find the earlist forked block in the unfinalised block array, which is currently in `getLastCorrectFinalizedBlock` that I think can be merged to `hasForked()`